database-couchdb/dataGenerator.js

```diff
@@ -0,0 +1,73 @@
+const { nano } = require('./index.js');
+const faker = require('faker');
+
+const totalReviews = 100000000;
+const batchSize = 1250;
+const simultaneousRequests = 5;
+
+const setupSeedDb = async () => {
+  try {
+    await nano.db.destroy('sdc')
+  } catch (err) {
+    console.log('Database does not exist')
+  }
+  const create = await nano.db.create('sdc');
+}
+
+const generateReviews = async () => {
+  let start = Date.now();
+  const db = await nano.db.use('sdc');
+  let documents = [];
+  let requests = [];
+  for (let i = 1; i <= totalReviews; i++) {
+    const review = {
+      reviewId: i,
+      title: faker.lorem.words(),
+      description: faker.lorem.paragraph(),
+      rating: Math.floor(Math.random() * (5 - 1 + 1) + 1),
+      isVerifiedPurchase: faker.random.boolean(),
+      isHelpfulCount: faker.random.number(),
+      reviewDate: faker.date.past(),
+      easeToUse: faker.random.number({ min: 1, max: 5 }),
+      voiceRecognition: faker.random.number({ min: 1, max: 5 }),
+      techSupport: faker.random.number({ min: 1, max: 5 }),
+      valueForMoney: faker.random.number({ min: 1, max: 5 }),
+      qualityOfMaterial: faker.random.number({ min: 1, max: 5 }),
+      batteryLife: faker.random.number({ min: 1, max: 5 }),
+      imageUrls: [
+        'http://placeimg.com/480/640/tech',
+        'http://placeimg.com/480/640/tech',
+        'http://placeimg.com/480/640/tech',
+        'http://placeimg.com/480/640/tech',
+        'http://placeimg.com/480/640/tech',
+      ],
+      productId: faker.random.number({ min: 1, max: 10000000 }),
+      color: faker.commerce.color(),
+      configuration: faker.lorem.words(),
+      customerId: faker.random.number({ min: 1, max: 10000000 }),
+      customerName: faker.name.findName(),
+      customerCountry: faker.address.country(),
+      profilePicUrl: 'http://placeimg.com/60/60/people',
+      category: 'Amazon Devices',
+      isBestSeller: faker.random.boolean(),
+    }
+    documents.push(review);
+    if (documents.length === batchSize) {
+      requests.push(db.bulk({ docs: documents }))
```

Comment:
can you explain this line to me?
here is what i have so far: you create a review and add it to an array. If the array length is equal to batch size, then you add the documents to the database and then push the "add documents operation" to the requests array. if requests array is of length 5, then you fire off all of the requests in batches of 5? what are the requests exactly? 

I believe you that this code works and it is well written, I just want to understand what is happening and the reason for it.

Reply:
So each document is just an object being held in memory. Once there's enough documents to meet the batch size, I fire off one batch upload (one http request which returns a promise) which is added to the requests array. Once the requests array has five promises in it, I make sure those five promises resolve before continuing to prevent storing too much in memory.